Library/OcAppleKernelLib/MkextContext.c

```diff
@@ -0,0 +1,1040 @@
+  http://opensource.org/licenses/bsd-license.php
+
+  THE PROGRAM IS DISTRIBUTED UNDER THE BSD LICENSE ON AN "AS IS" BASIS,
+  WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
+
+**/
+
+#include <Uefi.h>
+
+#include <IndustryStandard/AppleFatBinaryImage.h>
+
+
+#include <Library/BaseLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/DebugLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/OcAppleKernelLib.h>
+#include <Library/OcCompressionLib.h>
+#include <Library/OcFileLib.h>
+#include <Library/OcGuardLib.h>
+#include <Library/OcStringLib.h>
+
+#define MKEXT_OFFSET_STR_LEN    24
+
+//
+// Alignment to 8 bytes.
+//
+#define MKEXT_ALIGN(a)         (ALIGN_VALUE (a, sizeof (UINT64)))
+
+STATIC
+BOOLEAN
+ParseKextBinary (
+  IN OUT UINT8         **Buffer,
+  IN OUT UINT32        *BufferSize,
+  IN     BOOLEAN       Is64Bit
+  )
+{
+  EFI_STATUS        Status;
+  MACH_HEADER_ANY   *MachHeader;
+  
+  Status = FatFilterArchitectureByType (Buffer, BufferSize, Is64Bit ? MachCpuTypeX8664 : MachCpuTypeI386);
+  if (EFI_ERROR (Status)) {
+    return FALSE;
+  }
+  MachHeader = (MACH_HEADER_ANY *)* Buffer; // TODO alignment?
+
+  if ((!Is64Bit && MachHeader->Signature == MACH_HEADER_SIGNATURE)
+    || (Is64Bit && MachHeader->Signature == MACH_HEADER_64_SIGNATURE)) {
+    return TRUE;
+  }
+
+  return FALSE;
+}
+
+STATIC
+VOID
+UpdateMkextLengthChecksum (
+  IN MKEXT_HEADER_ANY   *Mkext,
+  IN UINT32             Length
+  )
+{
+  Mkext->Common.Length  = SwapBytes32 (Length);
+  Mkext->Common.Adler32 = SwapBytes32 (
+    Adler32 ((UINT8*)&Mkext->Common.Version,
+      Length - OFFSET_OF (MKEXT_CORE_HEADER, Version))
+      );
+}
+
+STATIC
+BOOLEAN
+ParseMkextV2Plist (
+  IN  MKEXT_V2_HEADER   *Mkext,
+  OUT UINT8             **Plist,
+  OUT XML_DOCUMENT      **PlistDoc,
+  OUT XML_NODE          **PlistBundles
+  )
+{
+  UINT8               *MkextBuffer;
+  UINT32              MkextLength;
+  UINT8               *PlistBuffer;
+  XML_DOCUMENT        *PlistXml;
+  UINT32              PlistOffset;
+  UINT32              PlistCompressedSize;
+  UINT32              PlistFullSize;
+  UINT32              PlistStoredSize;
+  UINT32              Tmp;
+
+  XML_NODE            *MkextInfoRoot;
+  UINT32              MkextInfoRootIndex;
+  UINT32              MkextInfoRootCount;
+
+  XML_NODE            *PlistBundleArray;
+  CONST CHAR8         *BundleArrayKey;
+  
+
+  MkextBuffer         = (UINT8 *) Mkext;
+  MkextLength         = SwapBytes32 (Mkext->Header.Length);
+  PlistOffset         = SwapBytes32 (Mkext->PlistOffset);
+  PlistCompressedSize = SwapBytes32 (Mkext->PlistCompressedSize);
+  PlistFullSize       = SwapBytes32 (Mkext->PlistFullSize);
+
+  PlistStoredSize = PlistCompressedSize;
+  if (PlistStoredSize == 0) {
+    PlistStoredSize = PlistFullSize;
+  }
+
+  if (OcOverflowAddU32 (PlistOffset, PlistStoredSize, &Tmp) || Tmp > MkextLength) {
```

Comment:
I do not see where does it check for plist being within the mkext file. This check is probably missing.